### PR TITLE
gh-108494: Argument clinic: Improve the `parse_file()` API

### PR DESCRIPTION
**Pull request by AlexWaygood**
- Get rid of the `ns` argument to `parse_file()`; restore the `verify` argument that was removed in https://github.com/python/cpython/commit/1dd951097728d735d46a602fc43285d35b7b32cb.
- Add a `limited_capi` argument to `parse_file()`. Make the `limited_capi` argument to both `Clinic.__init__` and `parse_file` _required_, so that callers always have to be explicit about whether the limited C API is desired or not. This means that there is a "single source of truth" about what the default is, rather than this being duplicated between the global constant and the logic in the `create_cli()` function.
- Get rid of `MockClinic` from the test file, which was added in https://github.com/python/cpython/commit/1dd951097728d735d46a602fc43285d35b7b32cb: just use our existing `_make_clinic()` helper function instead.

With this PR, we would need to make only these changes if we wanted to change it tomorrow so that the limited C API is the default for AC-generated code (plus a few fixes to some tests that currently depend on being able to use our internal C API):

```diff
diff --git a/Lib/test/test_clinic.py b/Lib/test/test_clinic.py
index efc6a922c5..f4f8b2dd20 100644
--- a/Lib/test/test_clinic.py
+++ b/Lib/test/test_clinic.py
@@ -20,7 +20,7 @@
     import clinic
     from clinic import DSLParser

-DEFAULT_LIMITED_C_API = False
+DEFAULT_LIMITED_C_API = True


 def _make_clinic(*, filename='clinic_tests'):
diff --git a/Tools/clinic/clinic.py b/Tools/clinic/clinic.py
index 2a4b6ff3b7..1a82ff7e57 100755
--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -6096,8 +6096,8 @@ def create_cli() -> argparse.ArgumentParser:
     cmdline.add_argument("--exclude", type=str, action="append",
                          help=("a file to exclude in --make mode; "
                                "can be given multiple times"))
-    cmdline.add_argument("--limited", dest="limited_capi", action='store_true',
-                         help="use the Limited C API")
+    cmdline.add_argument("--no-limited", dest="limited_capi", action='store_false',
+                         help="Don't use the Limited C API")
     cmdline.add_argument("filename", metavar="FILE", type=str, nargs="*",
                          help="the list of files to process")
     return cmdline
```

Closes #108504

* Issue: gh-108494

### CODE REVIEW

**Review by erlend-aasland — APPROVED** [review]

File: Lib/test/test_clinic.py [hunk 1]
```diff
@@ -13,25 +13,19 @@
...
     import clinic
     from clinic import DSLParser
 
-
-def default_namespace():
-    ns = types.SimpleNamespace()
-    ns.force = False
-    ns.limited_capi = clinic.DEFAULT_LIMITED_CAPI
-    return ns
+DEFAULT_LIMITED_C_API = False
```
Comment (vstinner) [inline Lib/test/test_clinic.py]: I added it to clinic.py to maybe change the default later. I don't think that it's relevant to have such constant in test_clinic. Just write False?

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -2415,7 +2414,7 @@ def __init__(
             *,
             filename: str,
             verify: bool = True,
-            limited_capi: bool = False,
+            limited_capi: bool
```
Comment (vstinner) [inline Tools/clinic/clinic.py]: It's kind of surprising for me to have a mandatory parameter after an optional parameter. Maybe move it before verify?

**Review by vstinner — APPROVED** [review]
LGTM, whatever works :-) I'm not going to nitpick on the coding style ;-)

File: Lib/test/test_clinic.py (hunk 1 above)
Comment (AlexWaygood) [reply Lib/test/test_clinic.py]: Hmm, yes, maybe this is over-engineered right now :)

**Comment by AlexWaygood** [comment]
Thanks for the reviews!

**Pull request merged**